src/libs/actions/Policy.ts

```diff
@@ -4091,6 +4122,245 @@ function setPolicyDistanceRatesDefaultCategory(policyID: string, currentCustomUn
     API.write(WRITE_COMMANDS.SET_POLICY_DISTANCE_RATES_DEFAULT_CATEGORY, params, {optimisticData, successData, failureData});
 }
 
+function prepareCustomUnitRatesArray(customUnitRates: Rate[]): Rate[] {
+    const customUnitRateArray: Rate[] = [];
+    customUnitRates.forEach((rate) => {
+        const cleanedRate = {...rate};
+        delete cleanedRate.pendingFields;
+        delete cleanedRate.errorFields;
+        customUnitRateArray.push(cleanedRate);
+    });
+
+    return customUnitRateArray;
+}
+
+function updatePolicyDistanceRateValue(policyID: string, customUnit: CustomUnit, customUnitRates: Rate[]) {
+    const currentRates = customUnit.rates;
+    const optimisticRates: Record<string, Rate> = {};
+    const successRates: Record<string, Rate> = {};
+    const failureRates: Record<string, Rate> = {};
+
+    Object.keys(customUnit.rates).forEach((rateID) => {
+        if (!customUnitRates.map((customUnitRate) => customUnitRate.customUnitRateID).includes(rateID)) {
+            return;
+        }
+
+        optimisticRates[rateID] = {
+            ...customUnitRates.find((rate) => rate.customUnitRateID === rateID),
+            pendingFields: {rate: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE},
+        };
+        successRates[rateID] = {
+            ...customUnitRates.find((rate) => rate.customUnitRateID === rateID),
+            pendingFields: {rate: null},
+        };
+        failureRates[rateID] = {
+            ...currentRates[rateID],
+            pendingFields: {rate: null},
+            errorFields: {rate: ErrorUtils.getMicroSecondOnyxError('common.genericErrorMessage')},
+        };
+    });
```

Comment:
I think we can optimize this a bit:

```suggestion
    const rateIDs = customUnitRates.map(rate => rate.customUnitRateID);
    
    for (const rateID of Object.keys(customUnit.rates)) {
        if (!rateIDs.includes(rateID)) {
            continue;
        }
        
        const foundRate = customUnitRates.find(rate => rate.customUnitRateID === rateID);
        optimisticRates[rateID] = {...foundRate, pendingFields: { rate: CONST.RED_BRICK_ROAD_PENDING_ACTION.UPDATE}};
        successRates[rateID] = { ...foundRate, pendingFields: { rate: null } };
        failureRates[rateID] = {
            ...customUnit.rates[rateID],
            pendingFields: { rate: null },
            errorFields: { rate: ErrorUtils.getMicroSecondOnyxError('common.genericErrorMessage') },
        };
    }
```